devito/ir/equations/algorithms.py

```diff
@@ -162,74 +164,139 @@ def concretize_subdims(exprs, **kwargs):
+    if d in mapper:
+        # Already have a substitution for this dimension
+        return
 
+    tkns = [tkn._rebuild(name=sregistry.make_name(prefix=tkn.name))
+            for tkn in d.tkns]
+    tkns_subs = {tkn0: tkn1 for tkn0, tkn1 in zip(d.tkns, tkns)}
+    left, right = [mM.subs(tkns_subs) for mM in (d.symbolic_min, d.symbolic_max)]
+    thickness = tuple((v, d._thickness_map[k]) for k, v in tkns_subs.items())
 
-@_concretize_subdims.register(ConditionalDimension)
-def _(d, mapper, sregistry):
-    # TODO: to be implemented as soon as we drop the counter machinery in
-    # Grid.__subdomain_finalize__
-    # TODO: call `_concretize_subdims(d.parent, mapper)` as the parent might be
-    # a SubDimension!
-    pass
+    mapper[d] = d._rebuild(symbolic_min=left, symbolic_max=right, thickness=thickness)
 
 
-@_concretize_subdims.register(MultiSubDimension)
-def _(d, mapper, sregistry):
-    if not d.is_abstract:
-        # TODO: for now Grid.__subdomain_finalize__ creates the thickness, but
-        # soon it will be done here instead
+@_concretize_subdims.register(ConditionalDimension)
+def _(d, mapper, rebuilt, sregistry):
+    if d in mapper:
+        # Already have a substitution for this dimension
         return
 
-    pd = d.parent
+    _concretize_subdims(d.parent, mapper, rebuilt, sregistry)
+
+    kwargs = {}
+
+    # Parent may be a subdimension
+    if d.parent in mapper:
+        kwargs['parent'] = mapper[d.parent]
+
+    # Condition may contain subdimensions
+    if d.condition is not None:
+        for v in d.condition.free_symbols:
+            _concretize_subdims(v, mapper, rebuilt, sregistry)
 
-    subs = mapper[pd]
+        if any(v in mapper for v in d.condition.free_symbols):
+            # Substitute into condition
+            kwargs['condition'] = d.condition.subs(mapper)
 
-    if d in subs:
+    if kwargs:
+        # Rebuild if parent or condition need replacing
+        mapper[d] = d._rebuild(**kwargs)
+
+
+@_concretize_subdims.register(MultiSubDimension)
+def _(d, mapper, rebuilt, sregistry):
+    if d in mapper:
         # Already have a substitution for this dimension
         return
 
-    name = sregistry.make_name(prefix=d.name)
-    ltkn, rtkn = MultiSubDimension._symbolic_thickness(name)
+    abstract_tkns = MultiSubDimension._symbolic_thickness(d.parent.name)
+    concrete_tkns = tuple(tkn._rebuild(name=sregistry.make_name(prefix=tkn.name))
+                          for tkn in abstract_tkns)
+
+    kwargs = {'thickness': concrete_tkns}
+    fkwargs = {}
+
+    idim0 = d.implicit_dimension
+    if idim0 is not None:
+        if idim0 in rebuilt:
+            idim1 = rebuilt[idim0]
+        else:
+            iname = sregistry.make_name(prefix='n')
+            rebuilt[idim0] = idim1 = idim0._rebuild(name=iname)
+
+        kwargs['implicit_dimension'] = idim1
+        fkwargs.update({'dimensions': (idim1,) + d.functions.dimensions[1:],
+                        'halo': None,
```

Comment:
I'm not sure these two Nones are actually necessary anymore?

Given https://github.com/devitocodes/devito/blob/master/devito/types/basic.py#L888-L889

Reply:
They are, but I've moved them to make it clear why they're needed (`function=None` is also set further down)